### PR TITLE
add lucene extension

### PR DESCRIPTION
**Pull request by yaochitc**
Removed the former pr and created a new one.The implementation is unfinished.

### CODE REVIEW

File: extensions-core/lucene-extensions/pom.xml [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+  <parent>
+    <groupId>io.druid</groupId>
+    <artifactId>druid</artifactId>
+    <version>0.9.1-SNAPSHOT</version>
+    <relativePath>../../pom.xml</relativePath>
+  </parent>
+
+  <properties>
+    <lucene.version>5.3.1</lucene.version>
```
Comment (javasoze) [inline extensions-core/lucene-extensions/pom.xml:38]: Upgrade to Lucene 6.0.0?

File: extensions-core/lucene-extensions/src/main/java/io/druid/lucene/LuceneDirectory.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package io.druid.lucene;
```
Comment (javasoze) [inline extensions-core/lucene-extensions/src/main/java/io/druid/lucene/LuceneDirectory.java:20]: Add license header

File: extensions-core/lucene-extensions/src/main/java/io/druid/lucene/LuceneDirectory.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ */
+public interface LuceneDirectory {
+    public int numRows();
+
+    public IndexReader getIndexReader() throws IOException;
```
Comment (javasoze) [inline extensions-core/lucene-extensions/src/main/java/io/druid/lucene/LuceneDirectory.java:30]: If user calls IndexReader.close on the returned IndexReader, and then when LuceneDirectory.close() is called, AlreadyClosedException will be thrown.

See comments on class ReadOnlyDirectory

File: extensions-core/lucene-extensions/src/main/java/io/druid/lucene/segment/loading/LuceneQueryableSegment.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
+package io.druid.lucene.segment.loading;
```
Comment (javasoze) [inline extensions-core/lucene-extensions/src/main/java/io/druid/lucene/segment/loading/LuceneQueryableSegment.java:20]: add Licensing header

File: extensions-core/lucene-extensions/src/main/java/io/druid/lucene/segment/loading/LuceneQueryableSegment.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import io.druid.segment.QueryableIndex;
+import io.druid.segment.StorageAdapter;
+import org.joda.time.Interval;
+
+import java.io.File;
+import java.io.IOException;
+
+/**
+ */
+public class LuceneQueryableSegment extends AbstractSegment {
```
Comment (javasoze) [inline extensions-core/lucene-extensions/src/main/java/io/druid/lucene/segment/loading/LuceneQueryableSegment.java]: Where is this class used?

File: extensions-core/lucene-extensions/src/main/java/io/druid/lucene/segment/loading/ReadOnlyDirectory.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    @Override
+    public int numRows() {
+        return indexReader.numDocs();
+    }
+
+    @Override
+    public IndexReader getIndexReader() {
+        return indexReader;
```
Comment (javasoze) [inline extensions-core/lucene-extensions/src/main/java/io/druid/lucene/segment/loading/ReadOnlyDirectory.java:46]: if caller calls IndexReader.close() on each thread, you will have failed other search threads with AlreadyClosedException.

The right pattern would be wrap IndexReader and on doClose(), call decRef instead. And call the actual close in the close() method. And for each getIndexReader() call, do incRef on the internal reader.

It is kinda tricky, let's have a test reproducing this scenario.

File: server/src/main/java/io/druid/guice/StorageNodeModule.java [hunk 7]
```diff
@@ -30,8 +30,8 @@
 import io.druid.query.DruidProcessingConfig;
 import io.druid.query.QueryRunnerFactoryConglomerate;
 import io.druid.segment.column.ColumnConfig;
-import io.druid.segment.loading.MMappedQueryableIndexFactory;
```
Comment (javasoze) [inline server/src/main/java/io/druid/guice/StorageNodeModule.java:33]: Let's leave class renaming to a different PR to minimize scope of this PR.

File: server/src/main/java/io/druid/segment/loading/MMappedQueryableSegmentFactory.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
+/*
```
Comment (javasoze) [inline server/src/main/java/io/druid/segment/loading/MMappedQueryableSegmentFactory.java:1]: Let's leave class renaming to a different PR to minimize scope of this PR.

File: server/src/main/java/io/druid/segment/loading/QueryableSegmentFactory.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
+/*
```
Comment (javasoze) [inline server/src/main/java/io/druid/segment/loading/QueryableSegmentFactory.java:1]: Let's leave class renaming to a different PR to minimize scope of this PR.

**Comment by javasoze** [comment]
Hi @yaochitc , please see my comments on the PR. Mainly 3 things:

1) cosmetic: e.g. licensing headers are missing with new files.
2) ReadOnlyDirectory.getIndexReader() and close() can be problematic.
3) Let's not do any file renaming for this PR, it is a better practice to have separate PRs for those.

Please see the my comments on the commits for details.

File: extensions-core/lucene-extensions/pom.xml (hunk 1 above)
Comment (yaochitc) [inline extensions-core/lucene-extensions/pom.xml:38]: This version is chosen as it is generally tested in other olap system. The newer version maybe not stable as it.

**Comment by yaochitc** [comment]
Thanks for commenting, I'll fix them.
